backend/libs/time-provider/src/fixed-time-provider.ts

```diff
@@ -0,0 +1,20 @@
+import { Injectable } from '@nestjs/common';
+import { TestTimeProvider } from './test-time-provider';
+import { TimeProvider } from '@coders-board-library/time-provider/time-provider';
+
+@Injectable()
+export class FixedTimeProvider implements TimeProvider, TestTimeProvider {
```

Comment:
Dziedziczenie po `TimeProvider` nie jest potrzebne. `TestTimeProvider` dziedziczy po `TimeProvider`.

Reply:
Niby tak, ale wtedy uzależniam się od szczegółów implementacyjnych TestTimeProvidera, akurat tak go zaimplementowałem. + Ważniejsze jest znaczenie semantyczne, a nie powtórzenie kodu. Czyli, ja chcę powiedzieć, że FiexedTimeProvider może być używany jako TimeProvider i też jako TestTimeProvider w testach. Jeśli tego nie zrobię, mój kod mówi tak jakbym chciał FixedTimeProvider używać jedynie jako TestTimeProvider.

Reply:
Problem tutaj jest trochę głębszy, bo nie powinno się rozszerzać funkcjonalności klasy na potrzeby testów. Po drugie metoda `moveCurrentDateTo(...)` jest funkcjonalnością specjalną dla `FixedTimeProvider`, a może i nawet dla `TimeProvider`.

Reply:
Hmm... to jakbyś to zrobił. Jak najbardziej się zgadzam, żeby nie rozszerzać klasy na potrzeby testów. Wtedy zrobiłbym sobie np. w Java implementację tylko dla testów, w osobnym source directory. W JavaScript widzę, że nawet kod testów jest jak najbliżej testowanych. 

Chodziło mi o to, żeby tylko w testach było można się poruszać w czasie, czyli 
```
export abstract class TestTimeProvider extends TimeProvider {
  abstract moveCurrentDateTo(date: Date);
}
```
Wszędzie indziej wykorzystujemy interfejs TimeProvider, a w testach TestTimeProvider - implementacja jedna. Nawet jakbym zrobił specjalną implementację TestTimeProvider załózmy jako klasę, to chyba w TS nie mogę wymusić jej użycia jedynie w testach?

Reply:
To czy kod testów będzie bliżej czy dalej funkcjonalności możemy ustalić między sobą, ja uważam, że powinien być jak najbliżej modułu, ale w osobnym folderze tests.

Jeśli chodzi o testowanie class zależnych od `TimeProvider` (inject-ujących `TimeProvider`) to w tym wypadku dajemy mocka `TimeProvider` i mamy przesuwanie czasu z poziomu [return value mocka](https://jestjs.io/docs/en/mock-functions.html#mock-return-values).

Natomiast jeśli klasa `FixedTimeProvider` ma mieć funkcjonalność modyfikacji czasu poza konstruktorem to powinna implementować odpowiedni interfejs np.:
```javascript
interface TimeProvider {
  currentDate(): Date;
}

interface MovableTimeProvider extends TimeProvider {
    moveCurrentDateTo(date: Date) : void;
}

export class FixedTimeProvider implements TimeProvider, MovableTimeProvider {
  constructor(private date: Date) {}

  currentDate(): Date {
    return this.date;
  }

  moveCurrentDateTo(date: Date) {
    this.date = date;
  }
}
```
`MovableTimeProvider` dziedziczy po `TimeProvider`, bo w sumie bez niego nie ma sensu ten interfejs. `FixedTimeProvider` implementuje obydwa z racji tego co napisałeś w poprzednim komentarzu 👍  tzn. `FixedTimeProvider` jest i `TimeProvider` i `MovableTimeProvider`. W przypadku zmiany interfejsu `MovableTimeProvider` nie psuje to implementacji dla klas używających `FixedTimeProvider` pod interfejsem `TimeProvider`. Testowane jest w tym wypadku można by rzec zbędne, ale 100% pokrycia jest zawsze lepsze niż 99% 😃 i polega na zmianie czasu (`moveCurrentDateTo`), a następnie sprawdzeniu czy `currentDate` zwraca zmieniony czas.

Reply:
To ja też jestem za tym. Jest osobny folder tests, ale na e2e. Proponuję tests i e2e-test (chociaż na razie takich nawet nie mamy i nie wiem czy warto).
Ale ładnie komentujesz z linkami :D 
Return value mocka znam, ale ogólnie jestem bardzo sceptyczny co do mocków. Często zmniejszają czytelnośc testów - np. zamiast mocków repo robię in-memory: https://youtu.be/ILBX9fa9aJo?t=3114
Ale w Javie nie są za wygodne i też mało wydajne, może to zboaczenie javowe :) 

FixedTimeProvider - nie ma mieć takiej funkcjonalności, więc OK. Jutro to przemyślę jeszcze :)

Reply:
A ja jestem zupełnie przeciwnego zdania tzn. że mocki poprawiają czytelność, ponieważ masz od razu explicit napisane jak się zachowuje mock, a nie musisz szukać w implementacji [stub-a](https://en.wikipedia.org/wiki/Test_stub). Co więcej pisząc stub-a musisz mieć pewność, że działa poprawnie tzn. powinieneś go przetestować. Oczywiście są sytuacje gdzie trzeba działanie czegoś zasymulować i łatwiej jest napisać stub niż mock. Zwykle dotyczy to jakiś zewnętrznych systemów lub podsystemów od których nasza implementacja jest zależna.

Reply:
OK, zgadzam się :) Zrobione.